docs/source/developers/index.md

```diff
@@ -228,6 +228,94 @@ class MyCompletionProvider(BaseProvider, FakeListLLM):
         )
 ```
 
+
+#### Using the full notebook content for completions
+
+The `InlineCompletionRequest` contains the `path` of the current document (file or notebook).
+Inline completion providers can use this path to extract the content of the notebook from the disk,
+however such content may be outdated if the user has not saved the notebook recently.
+
+The accuracy of the suggestions can be slightly improved by combining the potentially outdated content of previous/following cells
+with the `prefix` and `suffix` which describe the up-to-date state of the current cell (identified by `cell_id`).
+
+Still, reading the full notebook from the disk may be slow for larger notebooks, which conflicts with the low latency requirement of inline completion.
+
+A better approach is to use the live copy of the notebook document that is persisted on the jupyter-server when *collaborative* document models are enabled.
+Two packages need to be installed to access the collaborative models:
+- `jupyter-server-ydoc` (>= 1.0) stores the collaborative models in the jupyter-server on runtime
+- `jupyter-docprovider` (>= 1.0) reconfigures JupyterLab/Notebook to use the collaborative models
+
+Both packages are automatically installed with `jupyter-collaboration` (in v3.0 or newer), however installing `jupyter-collaboration` is not required to take advantage of *collaborative* models.
+
+The snippet below demonstrates how to retrieve the content of all cells of a given type from the in-memory copy of the collaborative model (without additional disk reads).
+
+```python
+from jupyter_ydoc import YNotebook
+
+
+class MyCompletionProvider(BaseProvider, FakeListLLM):
+    id = "my_provider"
+    name = "My Provider"
+    model_id_key = "model"
+    models = ["model_a"]
+
+    def __init__(self, **kwargs):
+        kwargs["responses"] = ["This fake response will not be used for completion"]
+        super().__init__(**kwargs)
+
+    def _get_prefix_and_suffix(self, request: InlineCompletionRequest):
+        prefix = request.prefix
+        suffix = request.suffix.strip()
+
+        server_ydoc = self.settings.get("jupyter_server_ydoc", None)
```

Comment:
The problem is that the provider does not have access to settings. We would need to expose the `jupyter_server_ydoc` either:
- on provider instances as an optional attribute,
- as an argument to `generate_inline_completions` and `stream_inline_completions`

if this is a use-case we want to support. Any thoughts?

Reply:
Alternatively, jupyter-ai could take the responsibility of extracting the document (`YNotebook`) and pass it to `generate_inline_completions`/`stream_inline_completions`. This way the providers would be restricted to accessing collaborative model of the document they were invoked on (rather than of any document user has open).

Reply:
Any thoughts @dlqqq?

Reply:
Checking back on this one :)

Reply:
If both solutions are fine and either would be accepted, that would be helpful to know too - more than radio silence. I just do not want to work on a PR to learn that the maintainers would not accept it anyways.

Reply:
@krassowski Hey Mike, sorry for missing this! I generally don't check in on draft PRs, as I thought the convention was that only open PRs are ready for review. Let me take a look.

Reply:
> The problem is that the provider does not have access to settings.

I think it makes sense to define `settings` as an optional instance attribute on `BaseProvider`, e.g. `settings: Optional[Dict[str, any]]`. There are other use-cases for chat language models for this as well, so this is a very reasonable addition. This attribute will be useless in the magics (as the server context is not available there), but I think that's fine. The type hint provided here should be sufficient to convey that info to devs. We can also include more details in a docstring for that attribute. 

> Alternatively, jupyter-ai could take the responsibility of extracting the document

I think we can explore this option later, but this `settings` API you are proposing should be implemented first to test completion with notebook context. If generated output turns out to be significantly better with notebook context, then I agree this makes sense to implement within Jupyter AI server extension itself rather than in the providers.

> `jupyter_server_ydoc`

I thought this package was renamed entirely to `jupyter_collaboration` after v1? IIRC `jupyter_server_ydoc` is essentially the 0.x major version of `jupyter_collaboration`.

Reply:
BTW, in the future, feel free to email me if you'd like to call my attention on anything! My email should be on my GitHub profile.

I get tons of GitHub notifications due to all the repos & orgs I'm in, and GitHub doesn't clearly distinguish between direct pings and random comments on old PRs/issues.

Reply:
> I thought this package was renamed entirely to `jupyter_collaboration` after v1?

Yes, it was.

> IIRC `jupyter_server_ydoc` is essentially the 0.x major version of `jupyter_collaboration`.

It was. Based on regular discussions in jupyter-server calls and in public issues, the `jupyter_server_ydoc` was resurrected in https://github.com/jupyterlab/jupyter-collaboration/pull/280 to allow for exactly this use case, as well as for using the server-side execution with shared model without forcing users to enable RTC. It is a part of the `jupyter-collaboration` monorepo, and pre-releases of version 1.0 are on PyPI, e.g. https://pypi.org/project/jupyter-server-ydoc/1.0.0a2/ (so `jupyter-collaboration` v3 will just depend on `jupyter_server_ydoc` along few others).

Reply:
Thank you for the context! @krassowski BTW, I will be away until Thursday, May 30 on personal leave. If you need our attention, please feel free to ping JasonWeill@, 3coins@, srdas@, or any other member of our team until I return.

Reply:
Some thoughts as I started an implementation (see the new commits below):
- should we call this `server_settings` to avoid confusion?
- should it be a class attribute? It seems easier from implementation perspective because then we can just set it once on `BaseProvider` level and then we do not need to update it ever; otherwise we have 4 places where providers get initialized so it would seem very repetitive
- using `MappingProxyType` over `Dict` would narrow down the public API
- ideally though I think we may want to filter which settings get exposed; this is not trivial - we cannot filter it at initialization time because extensions may load in different order so `jupyter_server_ydoc` may not yet be in the dictionary; instead it would need to be filtered on access (e.g. in a custom `__getitem__(self, key: str) -> Any`

Reply:
@krassowski Hey Mike, I just got back from vacation. Let me take a look at this either this afternoon or tomorrow morning.

Reply:
I hope you had a good vacation!

Reply:
@krassowski Got sick yesterday but feeling better. 

- I agree with using the `server_settings` attribute name.
- A class attribute is fine since this attribute should only be set by `jupyter_ai` once per Python process; I agree that using an instance attribute can be tedious because of how many places `BaseProvider` is being instantiated in.

> using MappingProxyType over Dict would narrow down the public API

Does this just serve to hint that this dictionary is read-only? I'm not familiar with when this type should be used.

> ideally though I think we may want to filter which settings get exposed; this is not trivial

I think it makes more sense to include the entire `server_settings` dictionary. I don't see an obvious benefit to only allow-listing some set of keys to `BaseProvider`. Is this motivated by a data privacy/security concern?

Reply:
>Does this just serve to hint that this dictionary is read-only? 

`MappingProxyType` is in fact read-only. Assigning to it would raise:

```
TypeError: 'mappingproxy' object does not support item assignment
```

> Is this motivated by a data privacy/security concern?

More that once we add things to API then we should not remove them without making a major release.